# 오차 (Error)
- 정확도의 함정

In [27]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

import pandas as pd
import numpy as np

In [28]:
cancer = load_breast_cancer()
# 위스콘신 주의 유방암 환자 데이터 셋
# 악성종양? 양성종양 여부
# target: 0: 악성종양, 1: 양성종양
# - WDBC-Malignant
# - WDBC-Benign

In [29]:
print(cancer['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [30]:
data = cancer['data']
target = cancer['target']
feature_names = cancer['feature_names']

In [31]:
df = pd.DataFrame(data=data, columns=feature_names)

In [34]:
df['target'] = cancer['target']

In [35]:
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [36]:
pos = df.loc[df['target'] == 1]
neg = df.loc[df['target'] == 0]

In [40]:
# 양성환자 수
pos 


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
19        13.540         14.36           87.46      566.3          0.09779   
20        13.080         15.71           85.63      520.0          0.10750   
21         9.504         12.44           60.34      273.9          0.10240   
37        13.030         18.42           82.61      523.8          0.08983   
46         8.196         16.84           51.71      201.9          0.08600   
..           ...           ...             ...        ...              ...   
558       14.590         22.68           96.39      657.1          0.08473   
559       11.510         23.93           74.52      403.5          0.09261   
560       14.050         27.15           91.38      600.4          0.09929   
561       11.200         29.37           70.67      386.0          0.07449   
568        7.760         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
19            0.08129         0.06664             0.047810         0.1885   
20            0.12700         0.04568             0.031100         0.1967   
21            0.06492         0.02956             0.020760         0.1815   
37            0.03766         0.02562             0.029230         0.1467   
46            0.05943         0.01588             0.005917         0.1769   
..                ...             ...                  ...            ...   
558           0.13300         0.10290             0.037360         0.1454   
559           0.10210         0.11120             0.041050         0.1388   
560           0.11260         0.04462             0.043040         0.1537   
561           0.03558         0.00000             0.000000         0.1060   
568           0.04362         0.00000             0.000000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
19                  0.05766  ...          19.26            99.70       711.2   
20                  0.06811  ...          20.49            96.09       630.5   
21                  0.06905  ...          15.66            65.13       314.9   
37                  0.05863  ...          22.81            84.46       545.9   
46                  0.06503  ...          21.96            57.26       242.2   
..                      ...  ...            ...              ...         ...   
558                 0.06147  ...          27.27           105.90       733.5   
559                 0.06570  ...          37.16            82.28       474.2   
560                 0.06171  ...          33.17           100.20       706.7   
561                 0.05502  ...          38.30            75.19       439.6   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     worst concave points  worst symmetry  worst fractal dimension  target  
19                0.12880          0.2977                  0.07259       1  
20                0.07283          0.3184                  0.08183       1  
21                0.06227          0.2450                  0.07773       1  
37                0.05013          0.1987                  0.06169       1  
46                0.02564          0.3105                  0.074

In [41]:
# 기존 양성 환자수 357에 악성환자 5개 추가
sample = pd.concat([pos, neg[:5]], sort=True)
sample

area error  compactness error  concave points error  concavity error  \
19      23.560           0.014600              0.013150          0.02387   
20      14.670           0.018980              0.006490          0.01698   
21      15.700           0.014320              0.014210          0.01985   
37      14.160           0.004899              0.011640          0.01343   
46       8.205           0.016460              0.005917          0.01588   
..         ...                ...                   ...              ...   
0      153.400           0.049040              0.015870          0.05373   
1       74.080           0.013080              0.013400          0.01860   
2       94.030           0.040060              0.020580          0.03832   
3       27.230           0.074580              0.018670          0.05661   
4       94.440           0.024610              0.018850          0.05688   

    fractal dimension error  mean area  mean compactness  mean concave points  \
19                 0.002300      566.3           0.08129             0.047810   
20                 0.002425      520.0           0.12700             0.031100   
21                 0.002968      273.9           0.06492             0.020760   
37                 0.001777      523.8           0.03766             0.029230   
46                 0.002582      201.9           0.05943             0.005917   
..                      ...        ...               ...                  ...   
0                  0.006193     1001.0           0.27760             0.147100   
1                  0.003532     1326.0           0.07864             0.070170   
2                  0.004571     1203.0           0.15990             0.127900   
3                  0.009208      386.1           0.28390             0.105200   
4                  0.005115     1297.0           0.13280             0.104300   

    mean concavity  mean fractal dimension  ...  worst area  \
19         0.06664                 0.05766  ...       711.2   
20         0.04568                 0.06811  ...       630.5   
21         0.02956                 0.06905  ...       314.9   
37         0.02562                 0.05863  ...       545.9   
46         0.01588                 0.06503  ...       242.2   
..             ...                     ...  ...         ...   
0          0.30010                 0.07871  ...      2019.0   
1          0.08690                 0.05667  ...      1956.0   
2          0.19740                 0.05999  ...      1709.0   
3          0.24140                 0.09744  ...       567.7   
4          0.19800                 0.05883  ...      1575.0   

    worst compactness  worst concave points  worst concavity  \
19            0.17730               0.12880          0.23900   
20            0.27760               0.07283          0.18900   
21            0.11480               0.06227          0.08867   
37            0.04619               0.05013          0.04833   
46            0.13570               0.02564          0.06880   
..                ...                   ...              ...   
0             0.66560               0.26540          0.71190   
1             0.18660               0.18600          0.24160   
2             0.42450               0.24300          0.45040   
3             0.86630               0.25750          0.68690   
4             0.20500               0.16250          0.40000   

    worst fractal dimension  worst perimeter  worst radius  worst smoothness  \
19                  0.07259            99.70        15.110           0.14400   
20                  0.08183            96.09        14.500           0.13120   
21                  0.07773            65.13        10.230           0.13240   
37                  0.06169            84.46        13.300           0.09701   
46                  0.07409            57.26         8.964           0.12970   
..                      ...              ...           ...               ...   
0                   0.11890     

In [43]:
# 학습 데이터 분할
x_train, x_test, y_train, y_test = train_test_split(sample.drop('target', 1), sample['target'], random_state=42)

In [44]:
from sklearn.linear_model import LogisticRegression 

In [45]:
model = LogisticRegression()
model.fit(x_train, y_train)
pred = model.predict(x_test)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [46]:
(pred == y_test).mean()

0.978021978021978

In [47]:
# 비교적 악성환자는 적기 때문에 다 1로 예측(양성으로)
my_prediction = np.ones(shape=y_test.shape)

In [49]:
# 정확도가 더 높다.
(my_prediction == y_test).mean()

0.989010989010989

정확도만 놓고 본다면 무조건 음성환자로 예측하는 분류기가 성능이 좋다.
하지만 의사가 무조건 **음성환자로 예측해서 예측률 98.9%로 말하는 의사**는 자질이 좋은 의사인가?

정확도만 보고 분류기의 성능을 판별하는 것은 **위와 같은 오류**가 있다.
이를 보완하기 위해, 정확도를 개선하기 위해 생겨난 지표들이 있다.